### PR TITLE
RichText: restore focus after setting selection

### LINKED ISSUE #17505: Changing a RichText value automatically focus the RichText
**Describe the bug**
Currently, if we change a RichText value the RichText gets focused and becomes the active element event if another element was focused and rich text value is changed using a programmatic interface e.g: `wp.data.dispatch( 'core/block-editor' ).updateBlockAttributes( block.clientId, { content: '1<strong>234</strong>56789' } );`.

Previously the popover was closed when a click outside them happened; now they are closed when the focus is outside them. This change makes the rich text automatically gaining focus problem worse. For example, if we have popover with a SelectControl that allows switching the content of a paragraph between some predefined values when the select is used, the popover closes.

Another side effect is that on PR https://github.com/WordPress/gutenberg/pull/16014 that adds custom color as a format, each time we try to use the color picker the popover closes, because when we use the color picker there is a change in the format leaving to a change to the content of paragraph and making the focus go outside the popover.
I guess we should address this issue before WordPress 5.3 because this problem may affect third-party plugins that add color options or even other option in a popover.

I guess we may also have a11y related problems with moving the focus.

**To reproduce**
- Insert paragraph block:
```
const block = wp.blocks.createBlock( 'core/paragraph', { content: '123456789' } );
wp.data.dispatch( 'core/block-editor' ).insertBlock( block );
```
- Put the cursor inside the paragraph e.g. in the middle of the paragraph.
Focus some element on the page (and verify it gets focused):
```
document.querySelector( 'button.editor-block-navigation' ).focus();
console.log( document.activeElement );
````
- Programatically change the paragraph content and verify the paragraph becomes the active element:
```
	wp.data.dispatch( 'core/block-editor' ).updateBlockAttributes( block.clientId, { content: '1<strong>234</strong>56789' } );
	console.log( document.activeElement );
```

Sep-20-2019 18-50-38

cc: @ellatrix

### PR DESCRIPTION
**Pull request by ellatrix**
## Description

Fixes #17505.

Attempt to NOT return focus to rich text after applying formatting, if focus has been moved elsewhere. This will now be the default for all buttons.

`RichText` will still allow focus to be set if it is desired, through `onChange( value, { focus: true } )`. This can be useful if you know that the currently focussed element will be removed from the DOM. This is the case for the link pop up UI.

## How has this been tested?

See #17505.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate.

### CODE REVIEW

**Comment by mcsf** [comment]
Whether we return or don't return focus in any given situation, this is generally a difficult problem full of nuance.

For instance, with accessibility in mind, should we make a distinction between clicking on a toolbar control (like Bold) and selecting it with the keyboard? If I select a portion of text and move my mouse to toggle Bold, then I may expect to resume typing, thus focus should be returned to RichText. On the other hand, if I select text, then work my way up to the toolbar controls using <kbd>Alt+F10</kbd> and some tabbing, then perhaps after toggling Bold with <kbd>Space</kbd> or <kbd>Enter</kbd> I would still like to tab to Italic or Link, especially since the toolbar is still visible. In that case, is my user expectation that I can still tab within the toolbar unreasonable? I have no idea!

**Comment by ellatrix** [comment]
@mcsf I couldn't agree more.

1. I definitely think that focus should remain in the toolbar if you tabbed all the way there. Perhaps to use other tools, or change your mind.
2. I also think that it would be good for focus to return to the text if you merely click the button... But this seems a more tricky situation to solve. Sometimes (the colour picker is a good example here) you don't want a mere click to return focus. In case of the link UI, there's also a lot more going on than just a click. Ideally, at the end, on submitting the link, I'd expect focus to go to the toolbar where you can view the link.

On the other hand, it kind of also makes sense that focus stays on the button, even if you merely clicked it. After all, you still clicked it, and therefore moved focus? I'm not sure about this one. Most editors do seem to return focus for a simple button click though.

Considering all this, it does seem best if rich text is not focussed after applying formatting, and to add exceptions from there.

It's also worth noting that **selection** is NOT **focus**. While we do not move focus, we still maintain the selection.

**Comment by ellatrix** [comment]
@mcsf We could built that special focus return behaviour into `RichTextToolbarButton`.

**Comment by mcsf** [comment]
> Most editors do seem to return focus for a simple button click though.
> […]
> It's also worth noting that **selection** is NOT **focus**. While we do not move focus, we still maintain the selection.

Yeah. I wonder if any editors handle this a bit more smartly, perhaps:
1. Whether with the mouse or keyboard, pressing into the Bold control does move focus to it.
1. Thus, any additional tabbing will move focus within the toolbar.
1. However, because selection remains on a portion of text, the editor knows that certain interactions — specifically, moving around with the keyboard arrows — are to be caught so that, then and only then, focus can be returned to the actual rich text.

**Comment by mcsf** [comment]
> @mcsf We could built that special focus return behaviour into `RichTextToolbarButton`.

Yeah, that sounds a bit in line with what I simultaneously commented. :)

**Comment by ellatrix** [comment]
Ok, I've modified `RichTextToolbarButton` to focus the last active element (rich text) if the button is clicked with the mouse. If the button is pressed otherwise, focus will remain on the button. I think this is good behaviour now.

File: packages/e2e-tests/specs/__snapshots__/rich-text.test.js.snap [hunk 1]
```diff
@@ -76,14 +82,14 @@ exports[`RichText should transform backtick to code 2`] = `
 <!-- /wp:paragraph -->"
 `;
 
-exports[`RichText should update internal selection after fresh focus 1`] = `
+exports[`RichText should undo backtick transform with backspace 1`] = `
```
Comment (ellatrix) [inline packages/e2e-tests/specs/__snapshots__/rich-text.test.js.snap]: These tests were not in alphabetic order... Jest reordered them.

**Comment by ellatrix** [comment]
This PR is ready for review now.

**Review by jorgefilipecosta — COMMENTED** [review]
Hi @ellatrix, this PR fixed the popover and seems an overall improvement 👍

I tested this PR together with advanced rich text plugin and I noticed the following issue not sure if related to this PR:
If we select some text and then go to the inline color format on the color popup we change the color to red using the input field, the color red is correctly applied, if we go to input field and type green the green is correctly applied, if then I type red again the color is not applied anymore and even if I change the color clicking on the visual color picker the color still does not applies.
Basically the color input works two times but stops working on the third.
Sep-27-2019 12-54-26

Another thing I noticed is that if I click on the bold or italic formats and then I press the arrow keys it moves the cursor on the paragraph but if I use code or subscript format then pressing right-left keys moves the focus on the toolbar buttons. I guess this is expected right given the new way we handle the focus?

**Comment by ellatrix** [comment]
@jorgefilipecosta Thanks for testing! It seems that focussing the input field will also move selection... a document can normally have not more than one selection. I need to think about how we can solve this... :/ Also now I see that for any button in the collapsed menu, focus won't be set back to rich text, but to the toggle button. That's because we look at the last focussed element. The only fix I think it to pass the rich text element to `RichTextToolbarButton`.

**Comment by ellatrix** [comment]
So, to recap the selection issue: I don't know how we can _both_ allow rich text selection and input field selection. 🤔

**Comment by jorgefilipecosta** [comment]
Testing PR https://github.com/WordPress/gutenberg/pull/16014 with this branch produces an even stranger dancing behavior:
Sep-27-2019 13-13-02


It seems like problems only happen when the format contains an input field.

**Comment by ellatrix** [comment]
Ok, found the issue. Working on a fix now.

**Comment by ellatrix** [comment]
@jorgefilipecosta Does the latest commit fix the issue with the colour picker? To me it looks good now.

**Comment by ellatrix** [comment]
I'll now try to fix the clicking for the buttons in the collapsed menu.

File: packages/format-library/src/link/inline.js [hunk 2]
```diff
@@ -149,9 +149,9 @@ class InlineLinkUI extends Component {
 
 		if ( isCollapsed( value ) && ! isActive ) {
 			const toInsert = applyFormat( create( { text: url } ), format, 0, url.length );
-			onChange( insert( value, toInsert ) );
+			onChange( insert( value, toInsert ), { focus: true } );
 		} else {
-			onChange( applyFormat( value, format ) );
+			onChange( applyFormat( value, format ), { focus: true } );
```
Comment (jorgefilipecosta) [inline packages/format-library/src/link/inline.js]: This API is interesting is it possible to call onChange without any change? So for example in the color picker when the popovers close and we know that we will not apply additional color changes we would call onChange to move the focus back to the rich text as the inline link does.

**Comment by jorgefilipecosta** [comment]
Hi @ellatrix, I found no issues with the color picker from "advanced rich text" and with the color picker being implemented in https://github.com/WordPress/gutenberg/pull/16014. Nice work 👍 

> Also now I see that for any button in the collapsed menu, focus won't be set back to rich text, but to the toggle button. That's because we look at the last focussed element.

Another side effect of this issue is that if we apply a code format and then a bold format, when we click the bold button the focus moves to the more richtext controls button and the tooltip becomes visible.

> The only fix I think it to pass the rich text element to RichTextToolbarButton.

This idea seems like a possible path; maybe a reference can be passed using context, so users of RichTextToolbarButton are not aware of that. Any undesirable side effects of following this path?

**Comment by ellatrix** [comment]
Thanks again for testing. Yeah the way I implemented it in the button is
flawed. Looking at the previous active element is not good, because in case
of the drop down buttons it would be the toggle button.

Thinking more about it, I actually think that we shouldn’t implement it at
the button level at all... we’re assuming that the button will apply a
format. A lot of buttons might not, and the core inline image button is a
perfect example.

So if we can’t do that? What’s left? We can track keyboard and mouse events
from the moment the rich text element blurs. E.g. if there’s only been
clicks, we set focus to rich text. But I’m not sure it this works either...
the color panel is also just clicks. So maybe after only a single click?
That again leaves out the drop down buttons...

So... I really don’t know how we can only move focus to rich text when
**clicking** a **RichTextButton** that does **not apply any formatting**.

Perhaps we should leave this out  and never return focus unless explicitly
requested? The only drawback is that, after clicking a button, you have to
click the text again to continue typing. Not great I guess.

I’ll think a bit more about it.

**Comment by jorgefilipecosta** [comment]
> Looking at the previous active element is not good, because in case
of the drop down buttons it would be the toggle button.

What if we never set the toggle button as the previous active element? When the more RichText Controls is the last focused moment we behave as we do if the last focused element was the richtext.

**Comment by paaljoachim** [comment]
Hey Ella @ellatrix How is this PR coming along?

It would be great to get this merged as the work on https://github.com/WordPress/gutenberg/pull/16014 and https://github.com/WordPress/gutenberg/issues/8171 will be able to continue.

Thanks!

**Comment by ellatrix** [comment]
New PR at #19536.

File: packages/format-library/src/link/inline.js (hunk 2 above)
Comment (ciprianimike) [reply packages/format-library/src/link/inline.js]: this is just what I was looking for, it works well. Thanks a lot for all your hard work @ellatrix @jorgefilipecosta
